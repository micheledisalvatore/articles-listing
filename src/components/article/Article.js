import React from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';

import { showArticle } from '../../actions/articles';

import { Content, Img, TextWrapper } from './Article.styled';

export const Article = ({ artwork, content, full, id, title, showArticleAction }) => {
  const handleArticleClick = () => { showArticleAction(id) };

  return (
    <div>
      <Img src={artwork} alt={title} />
      <TextWrapper>
        <h3>{title}</h3>
        <Content dangerouslySetInnerHTML={{__html: content}} onClick={handleArticleClick} full={full} />
      </TextWrapper>
    </div>
  )
}

Article.propTypes = {
  artwork: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  full: PropTypes.bool,
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  showArticleAction: PropTypes.func.isRequired,
}

Article.defaultProps = {
  full: false,
}

export const mapDispatchToProps = (dispatch) => ({
  showArticleAction: bindActionCreators(showArticle, dispatch),
})