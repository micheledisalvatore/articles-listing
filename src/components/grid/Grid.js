import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { values } from 'ramda';

import { getArticles } from '../../actions/articles';

import Article from '../article';

import { Wrapper } from './Grid.styled';

export class Grid extends Component {
  constructor(props) {
    super(props);

    const { getArticlesAction } = props;

    getArticlesAction();
  }

  render() {
    const { articles } = this.props;

    return (
      <Wrapper>
        { values(articles).map((article) => (
          <Article key={article.id} {...article} />
        ))}
      </Wrapper>
    )
  }
}

Grid.propTypes = {
  getArticlesAction: PropTypes.func.isRequired,
  articles: PropTypes.objectOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
    })
  ).isRequired,
}

export const mapStateToProps = ({ articles: { list } }) => ({
  articles: list,
})

export const mapDispatchToProps = (dispatch) => ({
  getArticlesAction: bindActionCreators(getArticles, dispatch),
});
