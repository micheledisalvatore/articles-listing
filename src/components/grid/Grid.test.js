/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';

import Article from '../article';

import { Grid, mapStateToProps } from './Grid';

describe('Given a Grid component', () => {
  let component;
  let wrapper;
  let getArticlesActionMock;

  describe('when it is rendered', () => {
    beforeEach(() => {
      getArticlesActionMock = jest.fn();
      const requiredProps = {
        getArticlesAction: getArticlesActionMock,
        articles: {},
      };
      component = <Grid {...requiredProps} />;
      wrapper = shallow(component);
    });

    afterEach(() => {
      getArticlesActionMock.mockReset();
    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call getArticlesActionMock', () => {
      expect(getArticlesActionMock).toHaveBeenCalledWith();
    });

    it('should NOT render any Article', () => {
      expect(wrapper.find(Article)).not.toExist();
    });

    describe('and a the articles are passed', () => {
      beforeEach(() => {
        wrapper.setProps({
          articles: {
            1: {
              id: 1,
            }
          },
        });
      });

      it('should match the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render 1 Article', () => {
        expect(wrapper.find(Article)).toHaveLength(1);
      });

      it('should render the Article with the passed id', () => {
        expect(wrapper.find(Article)).toHaveProp('id', 1);
      });
    });
  });
});

describe('Given a mapStateToProps function', () => {
  describe('when it is called', () => {
    it('should return an object with the passed values', () => {
      expect(mapStateToProps({
        articles: {
          list: {
            1: {
              id: 1,
              name: 'foo',
            },
          },
        },
      })).toEqual({
        articles: {
          1: {
            id: 1,
            name: 'foo',
          },
        },
      });
    });
  });
});
