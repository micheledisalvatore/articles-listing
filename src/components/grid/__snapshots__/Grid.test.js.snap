// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Grid component when it is rendered and a the articles are passed should match the snapshot 1`] = `
<styled.div>
  <Connect(Article)
    id={1}
    key="1"
  />
</styled.div>
`;

exports[`Given a Grid component when it is rendered should match the snapshot 1`] = `<styled.div />`;
